### PR TITLE
Gem doi

### PR DESCRIPTION
**Pull request by yemoski**
All the data of the gem datasets now have their doi showing on the front end

### CODE REVIEW

**Comment by nein09** [comment]
Nice. I liked your idea of building a list of prefixes to match and iterating through it better, though - it'll cut down on the constants that we have sitting around in the code, so please do that instead.

Incidentally, given the structure of g-e-m's metadata, I think we could also get a doi out of g-e-m datasets by changing our SPARQL query from doing

` ?s sschema:identifier | sschema:identifier/sschema:value ?id .`

to ` ?s sschema:identifier | sschema:identifier/sschema:url | sschema:identifier/sschema:value ?id .`

but I'm not sure which way would be faster.

**Comment by yemoski** [comment]
I just tried changing the query, I noticed that there were multiple results of the same dataset 
image
Should i leave the query as it was before

**Comment by nein09** [comment]
oh, right. Good question. It's finding both `url` and `value` in the identifier,  so it grabs both and treats them as two separate results.

We could do something like...

```?s { sschema:identifier } UNION {
     ?s sschema:identifier/sschema:url | sschema:identifier/sschema:value ?id . MINUS { sschema:identifier/sschema:url id . ?s sschema:identifier/sschema:value id .  }
}```

Maybe that's kind of gross though.

**Comment by yemoski** [comment]
I just tried to run the query on blazegraph
image

**Comment by yemoski** [comment]
I tried running the query on blazegraph
image

**Comment by nein09** [comment]
It might be that the `?s` goes inside the curly braces on line 47 there and it probably needs an `?id `. after `schema:identifier` too, so like `{ ?s schema:identifier ?id .} UNION ... `.

If that doesn't work out, let's not do it in the query. I was curious about how well it would go, but this doesn't need to turn into a rabbit hole.

**Comment by yemoski** [comment]
image
Thanks i tried that it didnt work, so i would revert back to the previous query i had

**Comment by yemoski** [comment]
The previous query seems to run very fast, I also wrote a script to check if we had multiple doi's and we don't
Happy days!!

**Comment by nein09** [comment]
Cool, can you still make this use a list of DOI prefixes that we're looking for and have it loop through that to call `startswith`? Thanks.

**Comment by yemoski** [comment]
added the prefix array

File: app/search/search.py [hunk 1]
```diff
@@ -35,13 +35,18 @@ def __init__(self, **kwargs):
...
-            if self.id and self.id.startswith('doi:'):
-                self.doi = self.id
-            elif any((match := url).startswith('http://dx.doi.org/') for url in self.urls):
-                self.doi = 'doi:' + match.lstrip('http://dx.doi.org/')
+            for x in prefixes:
+                if self.id and self.id.startswith(x):
+                    self.doi = self.id
+                elif any((match := url).startswith(x) for url in self.urls):
+                    self.doi = 'doi:' + match.lstrip(x)
+                elif any ((match := url).startswith(x) for url in self.urls):
```
Comment (nein09) [inline app/search/search.py:47]: Looks like you have a couple of duplicate lines here.

**Comment by yemoski** [comment]
Oh no!
I would hop on that now.

**Comment by yemoski** [comment]
This should work now, I added a break statement so it does not go through all the prefixes once it finds one of the prefix.
Should I write a test case for the gem datasets (http://data.g-e-m.dk/datasets?doi=) ?

**Comment by nein09** [comment]
Ah, both of those are good ideas, yes, please do write a test case for those.

**Comment by yemoski** [comment]
it looks good to be merged into main!

**Comment by nein09** [comment]
Looking good!

**Pull request merged**